### ISSUE TITLE
GZip attachments

### ISSUE THREAD

**Issue by random6886** [cloudant/nodejs-cloudant#72]
This might be already implemented, but I wasn't having any luck looking through here or nano. 

```
var fs = require('fs');

alice.attachment.get('rabbit', 'rabbit.png').pipe(fs.createWriteStream('rabbit.png'));
```

The above is an example for producing a stream, which I've been able to do successfully. Is there a way to have the stream comeback gzipped though? For example, if I do a:

```
curl https://(my_account).cloudant.com/db/doc/attachment --silent --write-out "%{size_download} - Took: %{time_total}s\n" --output /dev/null -u (my_account)
```

It produces: `2047050 - Took: 0.993s`

Now if I include the gzip/deflate header:

```
curl https://(my_account).cloudant.com/db/doc/attachment --silent --write-out "%{size_download} - Took: %{time_total}s\n" --output /dev/null -u (my_account) -H "Accept-encoding: gzip,deflate"
```

It produces: `134440 - Took: 0.383s`

Thanks!

**Comment by glynnbird** [comment]
This is a really good point. Doing `curl --compressed` should yield performance benefits by compressing the over-the-wire traffic, not just for attachments but for all substantial JSON transfers.

I've enabled it in the commit 6779d35 and we'll test it in the coming days, ahead of our 1.1 release, which is scheduled for 30th June. 

Many thanks for raising this!

**Comment by random6886** [comment]
I really appreciate how fast you responded to this issue! So would the data that I get back be automatically decompressed? 

The reason I ask is once my node application receives this data, I want to pass it on somewhere else (so would want to keep it compressed) until it reaches its final destination - say a client's browser.

Perhaps this could be a toggable setting to automatically decompress?

**Comment by glynnbird** [comment]
The data is automatically compressed and decompressed, so as a programmer, you see no difference. This is the documentation in the `request` module that is used under the hood:

```
gzip - If true, add an Accept-Encoding header to request compressed content encodings from the server (if not already present) and decode supported content encodings in the response. Note: Automatic decoding of the response content is performed on the body data returned through request (both through the request stream and passed to the callback function) but is not performed on the response stream (available from the response event) which is the unmodified http.IncomingMessage object which may contain compressed data. See example below.
```

**Comment by random6886** [comment]
Ah, so in my setup, I have:

**Website** ----> **Node.js Server** ----> **Cloudant**

When the user makes a request to the **Node.js server**, I request an attachment from **Cloudant** and then pipe it back to the **website** in the response. 

From what you're saying, under the hood, the request module would request a gzipped message from **Cloudant**, and then decompress it automatically on the **Node.js Server**?

I guess I have two problems:
1. I need a way for the response from **Cloudant** to be gzipped (performance), which I think you have just solved.
2. I need a way for the **Node.js Server** to **not** automatically decompress the **Cloudant** response because I would want to pipe the already gzipped message onto the **Website**

**Comment by glynnbird** [comment]
I don't know everything about your application, but I would say 

1. gzipping the output of your content (on its way to the browser) is computationally cheap and saves a ton of bandwidth. If you're using Express, then something like https://github.com/expressjs/compression makes it easy. It's a *little* inefficient to decompress & compress it again, but the advantages for the client are big.

2. If your Node.js Server is just proxying data between the browser and client, then you could consider just getting the client to communicate with Cloudant directly. With CORS enabled, your web app can make XHR requests to Cloudant without going through your Node.js servers at all. 

3. Doing some sanity checking, it seems to me that adding "Accept-encoding: gzip,deflate" only compresses attachments, not document bodies. I'm checking this with the Cloudant Dev team. If this is the case, then it's worth keeping (for attachments) but it isn't having the boost that you hoped it would.

I'll report back.

**Comment by random6886** [comment]
Thanks for the updates!

**Comment by glynnbird** [comment]
The official line on this is that adding "Accept-encoding: gzip,deflate"

* compresses some attachments (it depends on the mime type)
* incoming requests (client --> Cloudant) can be compressed and will be accepted - this is good news when adding and modifying data, especially on asymmetric connections (like ADSL)
* API responses are not compressed

I'm going to push for the third point to be pushed into the Dev backlog.

Meanwhile gzipping requests doesn't harm anything and has a positive impact on PUTs and POSTs, so I'll keep it in. If and when Cloudant supports gzipped responses, it'll 'just work'.

**Comment by jhs** [comment]
@random6886, thanks for your input! I think we are zeroing-in on a set of unit tests to make sure we have this working.

I think the question is, what "accept" headers are the clients sending? To my understanding, there are two clients: the browser accessing your Node server, and that Node server when accessing Cloudant. I should hope that, if your accept header is the same as your client's (i.e. either both say nothing, or both say "gzip"), then you should be able to stream the Cloudant attachment directly to the client, and compression should be on/off as indicated. If not, that is a bug in this library.

Boy, @glynnbird, I think a big theme of 1.2 is emerging: unit tests to actually confirm that messages on the wire are what we expect them to be. BTW, if you don't hear from the dev team, I strongly recommend Charles Proxy. It is a superb way to see what is up and down the wire. It has replaced tcpdump for me.

**Comment by glynnbird** [comment]
@jhs I think for this release, if we want to keep the gzip: true flag in place, then we need a test that ensures that it 

* doesn't break normal API interactions (as Cloudant isn't gzipping responses anyway)
* does allow attachments to be retrieved normally despite being gzipped
* does allow PUTs and POSTs to continue to work despite being gzipped

Personally I think the timescale is tight if we want to release on Tuesday, so I'm happy to retract the gzip:true flag for now and put it in the 1.2  list.

**Comment by jhs** [comment]
@glynnbird, yes all three points are correct. Just to clarify, when you say "despite being gzipped" I think you are exclusively talking about messages with `Transfer-Encoding: gzip, deflate` headers. (Correct me if I am wrong.)

I think we can do most or all of this testing with Nock. Nock can return binary data (a Node.js Buffer) so I think we can either save it in the test/ directory, or just encode it directly in the source. (It may be time for a new test file to reduce clutter.)

The only thing I'm not sure about Nock is, when we PUT/POST to the server, we should confirm that if we say the transfer is compressed, then it should be compressed, and vice versa. I don't think Nock can examine our payload very carefully. I think we can just fall back to the server you wrote in the test suite. It works fine, nobody has had any issues.

While I am brainstorming: for comprehensive testing, we would want to make sure that if we say gzip, we send gzipped data; if we say deflate, we send deflate data, and also that if we say we accept both, then either one does in fact work. To some extent that is overlapping with Nano and Request's job, but given that we had some issues already, do you think we should just test it ourselves?

**Comment by glynnbird** [comment]
@jhs My understanding is that when we supply "gzip:true" into requestDefaults, it adds `Accept-Encoding: gzip` (not deflate) into the headers and would automatically unzip any compressed replies. As Cloudant ignores this header at the moment and responds with uncompressed replies, there is no actual benefit. So my test only verifies that the header is sent, not that anything is compressed on the way up or down.

**Comment by jhs** [comment]
Ah! Well then omitting deflate makes things that much simpler.

**Comment by jhs** [comment]
Fixed in fb5c9e33
